Chap_API_Proc_Mgmt.tex

```diff
@@ -1285,3 +1320,345 @@ \subsection{\code{PMIx_Get_cpuset}}
+                          void *cbdata);
+\end{codepar}
+\cspecificend
+
+\begin{arglist}
+\argin{topo}{Pointer to the topology description of the node where the process is located (\code{NULL} indicates the local node) (\refstruct{pmix_topology_t})}
+\argin{cpuset}{Pointer to the location of the process (\refstruct{pmix_cpuset_t})}
+\argin{devices}{Bitmask identifying the type of devices whose distance is to be computed (\refstruct{pmix_device_type_t})}
+\argin{cbfunc}{Callback function \refapi{pmix_info_cbfunc_t} (function reference)}
+\argin{cbdata}{Data to be passed to the callback function (memory reference)}
+\end{arglist}
+
+Returns one of the following:
+
+\begin{itemize}
+\item \refconst{PMIX_SUCCESS} indicating that the request has been accepted for processing and the provided callback function will be executed upon completion of the operation. Note that the library must not invoke the callback function prior to returning from the \ac{API}.
+\item a non-zero \ac{PMIx} error constant indicating a reason for the request to have been rejected. In this case, the provided callback function will not be executed
+\end{itemize}
+
+
+%%%%
+\descr
+
+Non-blocking form of the \refapi{PMIx_Compute_distances} \ac{API}.
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+\subsection{Device Distance Callback Function}
+\declareapi{pmix_device_dist_cbfunc_t}
+
+%%%%
+\summary
+
+The \refapi{pmix_device_dist_cbfunc_t} is used to return an array of device distances.
+
+\versionMarker{4.0}
+\cspecificstart
+\begin{codepar}
+typedef void (*pmix_device_dist_cbfunc_t)
+    (pmix_status_t status,
+     pmix_device_distance_t *dist,
+     size_t ndist,
+     void *cbdata,
+     pmix_release_cbfunc_t release_fn,
+     void *release_cbdata);
+\end{codepar}
+\cspecificend
+
+\begin{arglist}
+\argin{status}{Status associated with the operation (\refstruct{pmix_status_t})}
+\argin{dist}{Array of \refstruct{pmix_device_distance_t} returned by the operation (pointer)}
+\argin{ndist}{Number of elements in the \argref{dist} array (\code{size_t})}
+\argin{cbdata}{Callback data passed to original \ac{API} call (memory reference)}
+\argin{release_fn}{Function to be called when done with the \argref{dist} data (function pointer)}
+\argin{release_cbdata}{Callback data to be passed to \argref{release_fn} (memory reference)}
+\end{arglist}
+
+
+%%%%
+\descr
+
+The \refarg{status} indicates if requested data was found or not.
+The array of \refstruct{pmix_device_distance_t} will contain the distance information.
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+\subsection{Device type}
+\declarestruct{pmix_device_type_t}
+
+The \refstruct{pmix_device_type_t} is a \code{uint64_t} bitmask for identifying the type(s) whose distances are being requested, or the type of a specific device being referenced (e.g., in a \refstruct{pmix_device_distance_t} object).
+
+\versionMarker{1.0}
+\cspecificstart
+\begin{codepar}
+typedef uint16_t pmix_device_type_t;
```

Comment:
I like where this is going, but have a suggestion.

What if a system has multiple types of GPUs, block devices, and networks - and we just want the distances to the specific subset? You kind of already hit this with network distinguishing between Ethernet vs OpenFabrics. I could also see wanting only the distances to the NVMe devices on the node and not the other storage devices for local checkpoint storage.

In the `endpoint_t` we return the `osname` of the device, maybe we can make this a slightly more complex type to incorporate something similar. Meaning:
```
typedef uint16_t pmix_device_type_class_t;

typedef struct pmix_device_type \{
    char *osname;
    pmix_device_type_class_t type;
\} pmix_device_type_t;
```

The `type_class_t` would be a single value for the class of the device. If `osname` is `NULL` then it works with all devices of that class, if `osname` is not NULL then that value is used to select the subset of devices that match that `osname` in this class. Maybe `osname` isn't the correct term, but you get the idea of what I'm talking about.
Then in the other APIs that consume this type they would take an array of this type.

For the network example, you would specify `{PMIX_DEVTYPE_NETWORK, "openfabrics"}` to get just the openfabrics devices. The value passed here is implementation-defined, but the keys (device classes) are standard-defined.

Actually, instead of adding a new data structure we could just add an attribute to the APIs with standard-defined keys for the device class, and values defined by the implementation. Then the APIs just need an info array argument to specify the device target(s).

Reply:
I confess I had to re-read this several times to really grok it, but I think I caught the gist. I have updated the `PMIx_Compute_distances` APIs to take an array of info instead of the `pmix_device_type_t` argument so we have more flexibility. See what you think.

Reply:
I think that's fine. I guess I was thinking that we could eliminate `pmix_device_type_t `, convert these constants to attributes, and then use some more generic attribute/value pairs to specify the device type.

So we would have the attributes:
 * `PMIX_DEVTYPE_BLOCK` with values like "sda,sdb" or "nvme"
 * `PMIX_DEVTYPE_COPROC` with values that are implementation-specific for "gpu", "mic". The user could be even more specific if they just want a subset of the devices "mic0,mic1"
 * `PMIX_DEVTYPE_NETWORK` with values like "ethernet", "openfabrics". Or more specific "eth0,eth1" or "mlx5_0,mlx5_1"

Then the set of attributes can be a bit more general, but the values can as specific as the user needs them to be. They can specify multiple attributes in the info keys if they want multiple device types. Instead of a comma-separated value like "eth0,eth1" we could also say that the value is a single item, and if they want to specify more than one value they duplicate the attribute in the info key array.

Reply:
I was trying to avoid the "magic word" problem - e.g., making users "know" that the right word is "ethernet" for the attribute value to get what they want. If we move to what you propose, it seemed to me that we also then need a list of string constants (e.g., `PMIX_OPENFABRICS` = "pmix.openfab") to help people specify something the implementation can understand. Once you take that step, you're pretty much right back to this PR, with string constants instead of numerical ones.

Reply:
That's a fair point. (Similar to what the storage folks were talking about today) I'm ok with leaving this like it is here.